### PR TITLE
Adding adjustable time range to single stock chart

### PR DESCRIPTION
**Pull request by kamalq97**
Changelog:
- In **urls.py** file, added new entry for historic single stock with a defined time range
- Defined custom `InvalidTimeRange` exception and used it in **stock_api.py**
- In **views.py** file, added optional `time_range` parameter with a default value of '1m' (one month)
- In **single_stock.html**, added tabs to show different time scales and the line chart color also changes (to red or green) depending on stock performance 
- Added time range unit tests to **test_stock_api.py**
- In **test_views.py**, one endpoint has been removed since it now triggers a different error response: 400 [bad request] for invalid time range instead of the previous 404 [not found]

### CODE REVIEW

File: myapp/templates/single_stock.html [hunk 1]
```diff
@@ -61,64 +61,81 @@
...
-			<a class="nav-link" data-toggle="pill" href="#max" role="tab" aria-controls="pills-contact" aria-selected="false">Max</a>
-		  </li>
-		</ul>
-		-->
+        <div id ="chartContainer">
+            <canvas id="stockChart"></canvas>
+        </div>
+        <ul class="nav nav-pills mb-3" id="pills-tab" role="tablist">
+            <li class="nav-item">
+                <a class="nav-link" data-toggle="pill" onclick="getHistoricData('1d')" role="tab" aria-controls="pills-home" aria-selected="true">1D</a>
```
Comment (murradkh) [inline myapp/templates/single_stock.html]: you used AjAX here!  Good one!

File: myapp/templates/single_stock.html [hunk 2]
```diff
@@ -61,64 +61,81 @@
...
-						}
-					]
-				},
-				options: {
-					responsive: true
-				}
-			});
-		});
-    });
+    $(document).ready(getHistoricData());
```
Comment (murradkh) [inline myapp/templates/single_stock.html:99]: small note, the JavaScript getting bigger here in this file, preferred to be separated and to include into a new JSscript file in `static/js/` folder

File: myapp/tests/test_stock_api.py [hunk 3]
```diff
@@ -10,35 +10,55 @@ def setUp(self):
...
-        self.assertEquals(second_month - first_month, 1)
+
+    def test_get_stock_historic_prices_time(self):
+        for time_range in self.valid_time_ranges:
+            response = get_stock_historic_prices(self.existed_symbols[0], time_range)
+
+            if time_range.endswith('d'):
+                first_day = datetime.fromisoformat(response[0].get('date')).day
+                last_day = datetime.fromisoformat(response[-1].get('date')).day
+                self.assertEquals(last_day - first_day, int(time_range[0])-1)
```
Comment (murradkh) [inline myapp/tests/test_stock_api.py]: Does this handle the issue you mentioned about Friday and Saturday having no stocks-reports?

**Review by murradkh — COMMENTED** [review]
Good Job!
with respect to the client-side view, two small comments:
1)  the buttons don't look like a clickable option (to act as a button)
2) when we click on 1D / 1Y/ max, will look like a messy data representation, hard to read, and many unneeded data/points. what we need I think for example when clicking on "1Y", to see only points on each month, and if I need more information I pick smaller time-range options.
so if we can fix this would be great.

File: myapp/templates/single_stock.html [hunk 4]
```diff
@@ -122,9 +138,13 @@
                             }
                         ]
                     },
-                    options: {responsive: true}
-                });
-            }
+                    options: 
+                        {
+                            responsive: true,
+                            spanGaps: true,
```
Comment (murradkh) [inline myapp/templates/single_stock.html:144]: Nice! you fixed the null points In an easy way

File: myapp/templates/single_stock.html [hunk 5]
```diff
@@ -61,64 +61,101 @@
...
+    }
+
+    function getGraphColors(p1, p2) {
+        if (p2 >= p1) {
+            return ['rgba(2, 144, 62, .2)', 'rgba(2, 144, 62, .5)']
+        }
+        else {
+            return ['rgba(220, 24, 24, .2)', 'rgba(220, 24, 24, .5)']
+        }
+    }
```
Comment (murradkh) [inline myapp/templates/single_stock.html]: you can use random color to be returned whenever it's called, like this (copied):

> 
> function random_rgba() {
>     var o = Math.round, r = Math.random, s = 255;
>     return 'rgba(' + o(r()*s) + ',' + o(r()*s) + ',' + o(r()*s) + ',' + r().toFixed(1) + ')';
> }

I also would use this for stocks-comparison.

**Review by murradkh — APPROVED** [review]
working fine! good job  شغيل 👍 
I see that's you have many changes which overlap with mine in #33, many conflicts expected. so a lot of work will be. that's should not happen in general I think, we did a mistake by start working on features overlapping in code places.
next time we need to avoid this mistake.
for the current state, what I propose to do is that after you merge your PR, I will try to resolve the conflicts, and make sure my features are working properly, then you make sure your features are working properly (on the stocks-comparison branch)
what do you think?

File: myapp/templates/single_stock.html (hunk 5 above)
Comment (kamalq97) [reply myapp/templates/single_stock.html]: Okay, I will add this to my code

**Pull request merged**